Java: basic version of automodel extraction query docs (#14487)
github/codeql: java/ql/automodel/src/README.md

# Automodel Java Extraction Queries

> This pack contains the automodel extraction queries for Java.

Reviewer: I think this document is missing a one or two sentence description of why someone would want to run them. Also worth mentioning that these queries are meant to be run as part of automated tooling. Since the tooling is not yet publicly released, we don't need to go into too much detail, though.

Author: done! thank you :) cffcc7334dc70d6207a590eb0a4001fa763e2c27

## Extraction Queries in `java/ql/automodel/src`

> This pack contains extraction queries for application mode and framework mode.

Reviewer: Minor: this sentence is half-overlapping with the previous sentence. 

```suggestion
Included in this pack are queries for both application mode and framework mode.
```

| Kind | Mode | Query File |
|------|------|------------|
| Candidates | Application Mode | [AutomodelApplicationModeExtractCandidates.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeExtractCandidates.ql) |

> | + Examples | Application Mode | [AutomodelApplicationModeExtractPositiveExamples.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeExtractPositiveExamples.ql) |

Reviewer: Minor: I find `+ Examples` slightly hard to grok. I know you don't want the columns too wide. Can you do `Pos Examples` and `Neg Examples`? Or add a newline between `Postive` and `Examples`?

Author: I've been using this term, as we often use this in written conversations. In this case, we have the space — I've replaced the +/- by positive/negative throughout the document!

| - Examples | Application Mode | [AutomodelApplicationModeExtractNegativeExamples.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeExtractNegativeExamples.ql) |
| Candidates | Framework Mode | [AutomodelFrameworkModeExtractCandidates.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelFrameworkModeExtractCandidates.ql) |
| + Examples | Framework Mode | [AutomodelFrameworkModeExtractPositiveExamples.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelFrameworkModeExtractPositiveExamples.ql) |
| - Examples | Framework Mode | [AutomodelFrameworkModeExtractNegativeExamples.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelFrameworkModeExtractNegativeExamples.ql) |

## Running the Queries

> The extraction queries are part of a separate query pack, `java-automodel-queries`. Use this pack to run them. The queries are tagged appropriately, you can use the tags (example here: https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeExtractNegativeExamples.ql#L8) to construct query suites.

Reviewer: ```suggestion
The extraction queries are part of a separate query pack, `codeql/java-automodel-queries`. Use this pack to run them. The queries are tagged appropriately, you can use the tags (example here: https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeExtractNegativeExamples.ql#L8) to construct query suites.
```

For example, a query suite selecting all example extraction queries (positive and negative) for application mode looks like this:

```
# File: automodel-application-mode-extraction-examples.qls
# ---
# Query suite for extracting examples for automodel

- description: Automodel application mode examples extraction.
- queries: .
  from: codeql/java-automodel-queries
- include:
    tags contain all:
    - automodel
    - extract
    - application-mode
    - examples
```

## Important Software Design Concepts and Goals

### Concept: `Endpoint`

> Endpoints are source code locations of interest. All +/- examples and all candidates are endpoints, but not all endpoints are examples or candidates. Each mode decides what endpoints are relevant. For instance, if the Java application mode wants to support candidates for sinks that are arguments passed to unknown method calls, then the Java application mode implementation needs to make sure that method arguments are endpoints. If you look at the `TApplicationModeEndpoint` implementation in [AutomodelApplicationModeCharacteristics.qll](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeCharacteristics.qll), you can see that this is the case: the `TExplicitArgument` implements this behavior.

Reviewer: Minor:

```suggestion
Endpoints are source code locations of interest. All +/- examples and all candidates are endpoints, but not all endpoints are examples or candidates. Each mode decides which endpoints are relevant. For instance, if the Java application mode wants to support candidates for sinks that are arguments passed to unknown method calls, then the Java application mode implementation needs to make sure that method arguments are endpoints. If you look at the `TApplicationModeEndpoint` implementation in [AutomodelApplicationModeCharacteristics.qll](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelApplicationModeCharacteristics.qll), you can see that this is the case: the `TExplicitArgument` implements this behavior.
```

Whether or not an endpoint is a +/- example, or a candidate depends on the individual extraction queries.

### Concept: `EndpointCharacteristics`

In the file [AutomodelSharedCharacteristics.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelSharedCharacteristics.ql), you will find the definition of the QL class `EndpointCharacteristic`.

An endpoint characteristic is a QL class that "tags" all endpoints for which the characteristic's `appliesToEndpoint` predicate holds. The characteristic defines a `hasImplications` predicate that declares whether all the endpoints should be considered as sinks/sources/negatives, and with which confidence.

> The +/- and candidate extraction queries largely<sup>[1](#largely-use-characteristics)</sup> use characteristics to decide which endpoint to select. For instance, if a characteristic exists that applies to an endpoint, and the characteristic implies (cf. `hasImplications`) that the endpoint is a sink with a high confidence &ndash; then that endpoint will be selected as a positive example. See the use of `isKnownAs` in [AutomodelFrameworkModeExtractPositiveExamples.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelFrameworkModeExtractPositiveExamples.ql).

Reviewer: Minor: There is a more canonical way of doing footnotes. See https://www.markdownguide.org/extended-syntax/#footnotes. Formatting is nicer, but this will move the footnote to the end of the page. You can decide which is more important.

Author: Oh, that's nice! Thank you! 0f2db1bcdba0dd3e0f44732c64f318d1a594a80a

<a name="largely-use-characteristics">1</a>: Candidate extraction queries are an exception, they treat `UninterestingToModelCharacteristic` differently.

#### :warning: Warning

> Do not to "fix" shortcomings that could be fixed by a better prompt or better example selection by adding language- or mode-specific characteristics . Those "fixes" tend to be confusing downstream when questions like "why wasn't this location selected as a candidate?" becomes progressively harder and harder to answer. It's best to rely on characteristics in the code that is shared across all languages and modes (see [Shared Code](#shared-code)).

Reviewer: ```suggestion
Do not try to "fix" shortcomings that could be fixed by a better prompt or better example selection by adding language- or mode-specific characteristics . Those "fixes" tend to be confusing downstream when questions like "why wasn't this location selected as a candidate?" becomes progressively harder and harder to answer. It's best to rely on characteristics in the code that is shared across all languages and modes (see [Shared Code](#shared-code)).
```

## Candidate Examples

This section contains a few examples of the kinds of candidates that our queries might select, and why.

:warning: For clarity, this section presents "candidates" that are **actual** sinks. Therefore, the candidates presented here would actually be selected as positive examples in practice - rather than as candidates.

### Framework Mode Candidates

Framework mode is special because in framework mode, we extract candidates (as well as examples) from the implementation of a framework or library while the resulting models are applied in code bases that are _using_ the framework or library.

In framework mode, endpoints currently can have a number of shapes (see: `newtype TFrameworkModeEndpoint` in [AutomodelApplicationModeExtractCandidates.ql](https://github.com/github/codeql/blob/main/java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll)). Depending on what kind of endpoint it is, the candidate is a candidate for one or several extensible types (eg., `sinkModel`, `sourceModel`).

#### Framework Mode Sink Candidates

> Sink candidates in framework mode are inputs to calls. As, in framework mode, we work on the implementation of a callable, these inputs are represented by a method's parameter definition.

Reviewer 2: The first sentence here is confusing to me, since we're looking at the framework itself the function calls are not present.
```suggestion
Sink candidates in framework mode are modelled as formal parameters of functions defined within the framework. We use these to represent the corresponding inputs of function calls in a client codebase, which would be passed into those parameters.
```

For example, customer code could call the `Files.copy` method:

```java
// customer code using a library
...
Files.copy(userInputPath, outStream);
...
```

In order for `userInputPath` to be modeled as a sink, the corresponding parameter must be selected as a candidate. In the following example, assuming they're not modeled yet, the parameters `source` and `out` would be candidates:

```java
// Files.java
// library code that's analyzed in framework mode
public class Files {
  public static void copy(Path source, OutputStream out) throws IOException {
    // ...
  }
}
```

#### Framework Mode Source Candidates

Source candidates are a bit more varied than sink candidates:

##### Parameters as Source Candidates

> A parameter could be a source, eg. when a framework passes user-controlled data to a handler defined in customer code.

Reviewer 2: Good explanation. I see how we want parameters as both source and sink candidates, though in practice the probability of the _same_ parameter being both is I think slim.

Reviewer: Minor

```suggestion
A parameter could be a source, e.g. when a framework passes user-controlled data to a handler defined in customer code.
```
